File: Daily-Streak/Day-13_Separate-Squares-I/Solution.py
class Solution:
    def separateSquares(self, squares: List[List[int]]) -> float:
        events = []
        total_area = 0.0

        for _, y, side in squares:
            total_area += side * side
            events.append((y, side))
            events.append((y + side, -side))

        events.sort()

        target = total_area / 2.0
        curr_area = 0.0
        curr_width = 0.0

        for i in range(len(events) - 1):
            curr_width += events[i][1]
            y1 = events[i][0]
            y2 = events[i + 1][0]

            if curr_width <= 0:
                continue

            strip_area = (y2 - y1) * curr_width

            if curr_area + strip_area >= target:
                remaining = target - curr_area
                return y1 + remaining / curr_width

            curr_area += strip_area

        return 0.0

        """
        ===========================
        ❌ WRONG LOGIC (COMMENTED)
        ===========================
        - Summing square areas directly
        - Overlaps break correctness
        """

File: Daily-Streak/Day-13_Separate-Squares-I/Solution.cpp
class Solution {
public:
    double separateSquares(vector<vector<int>>& squares) {
        // Each event: (y, delta_width)
        vector<pair<double, double>> events;
        double totalArea = 0.0;

        for (auto &sq : squares) {
            double y = sq[1];
            double side = sq[2];
            totalArea += side * side;

            events.push_back({y, +side});         // square starts
            events.push_back({y + side, -side});  // square ends
        }

        sort(events.begin(), events.end());

        double target = totalArea / 2.0;
        double currArea = 0.0;
        double currWidth = 0.0;

        for (int i = 0; i + 1 < events.size(); i++) {
            currWidth += events[i].second;
            double y1 = events[i].first;
            double y2 = events[i + 1].first;

            if (currWidth <= 0) continue;

            double stripArea = (y2 - y1) * currWidth;

            if (currArea + stripArea >= target) {
                double remaining = target - currArea;
                return y1 + remaining / currWidth;
            }

            currArea += stripArea;
        }

        return 0.0;
    }

    /*
    ===========================
    ❌ WRONG IDEA (COMMENTED)
    ===========================
    - Adding square areas directly
    - Ignores vertical overlaps
    */
};
